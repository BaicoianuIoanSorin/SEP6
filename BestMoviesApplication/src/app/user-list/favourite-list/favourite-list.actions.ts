export class UserFavouriteListMoviesFetch {
  static readonly type = 'Favourite list - Fetching info';
  constructor(public userName: string) {
  }
}
export class UserFavouriteListMoviesReset {
  static readonly type = 'Favourite list - reset';
  constructor() {
  }
}
