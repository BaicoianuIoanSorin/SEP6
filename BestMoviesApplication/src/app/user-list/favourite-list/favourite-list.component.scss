nb-card-header {
  display: flex;
  justify-content: space-between;
  align-items: center;
  .title {
    flex-grow: 1;
    text-align: center;
  }

  .right-side {
    display: inline-flex;
    gap: 10px;
  }
}

.favourite-list {
  .movie-container {
    display: grid;
    grid-template-columns: repeat(auto-fill, minmax(200px, 1fr));
    grid-gap: 1rem;
    max-width: 1200px;
    margin: 0 auto;

    .poster-movie {
      position: relative;
      width: 100%;
      border-radius: 15px 15px 15px 15px;
      box-shadow: 8px 6px 20px 2px #6d6d6d;

      .header-poster-movie {
        .movie-title {
          height: 2.8em;
        }
      }

      .body-poster-movie {
        padding: unset;
        overflow: hidden;
        position: relative;

        .show-more-button {
          position: absolute;
          top: 50%;
          left: 50%;
          transform: translate(-50%, -50%);
          opacity: 0;
          transition: opacity 0.3s ease;
          display: flex;
          align-items: center;
          justify-content: center;
          z-index: 1; /* Ensure the button is above the image */
        }

        img {
          width: 100%;
          height: 100%;
          object-fit: cover;
          object-position: center;
        }
      }

      .footer-poster-movie {
        margin-top: 0px;

        .movie-year {
          margin-top: 1rem;
        }
      }

    }

    .poster-movie:hover {
      opacity: 0.6;
      cursor: pointer;
    }

    .poster-movie:hover .body-poster-movie .show-more-button {
      opacity: 1;
    }
  }

  .best-movies {
    width: 100%;
  }
}

.spinner {
  color: nb-theme(color-primary-500);
}
