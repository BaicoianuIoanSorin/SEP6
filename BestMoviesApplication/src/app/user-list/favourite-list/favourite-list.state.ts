import { Movie } from "src/model/movie";
import { Action, State, StateContext } from "@ngxs/store";
import { Injectable } from "@angular/core";
import { NbToastrService } from "@nebular/theme";
import { produce } from "immer";
import {
  UserFavouriteListMoviesRemoveMovieFromFavourites,
  UserFavouriteListMoviesFetch,
  UserFavouriteListMoviesReset, UserFavouriteListMoviesResetFiltering
} from "src/app/user-list/favourite-list/favourite-list.actions";
import { catchError, tap } from "rxjs/operators";
import { throwError } from "rxjs";
import { FavoritesService } from "src/api/favorites.service";
import {HttpErrorResponse} from "@angular/common/http";

export interface UserFavouriteListMoviesStateModel {
  isFetching: boolean;
  isFiltered: boolean;
  movies: Movie[];
}

export const defaultsState: UserFavouriteListMoviesStateModel = {
  isFetching: false,
  isFiltered: false,
  movies: [],
}

@State<UserFavouriteListMoviesStateModel>({
  name: 'userFavouriteListMoviesPage',
  defaults: defaultsState,
})

@Injectable()
export class UserFavouriteListMoviesState {

  movies: Movie[] = [];
  constructor(
    private nbToastrService: NbToastrService,
    private favoritesService: FavoritesService,
  ) {
  }

  @Action(UserFavouriteListMoviesFetch)
  async userFavouriteListMoviesFetch(
    {getState, setState}: StateContext<UserFavouriteListMoviesStateModel>,
    action: UserFavouriteListMoviesFetch) {

    let newState = produce(getState(), draft => {
      draft.isFetching = true;
    })
    setState(newState);

    return this.favoritesService.getFavorites(action.userName)
    .pipe(
      tap((movies: Movie[]) => {
        newState = produce(getState(), draft => {
          draft.movies = movies;
          if(action.isFilteringAction) {
            draft.isFiltered = true;
            console.log("xd")
          }
          else {
            this.movies = movies;
          }
          draft.isFetching = false;
        });
        setState(newState);
      }),
      catchError((error) => {
        newState = produce(getState(), draft => {
          draft.isFetching = false;
        });
        this.nbToastrService.show("API error", "Could not fetch top list of movies for you", {status: "danger"})
        setState(newState);
        return throwError(error);
      })
    );
  }

  @Action(UserFavouriteListMoviesRemoveMovieFromFavourites)
  userFavouriteListMoviesAddMovieToFavourites(
    {getState, setState}: StateContext<UserFavouriteListMoviesStateModel>,
    action: UserFavouriteListMoviesRemoveMovieFromFavourites) {

    let newState = produce(getState(), draft => {
      draft.isFetching = true;
    })
    setState(newState);

    this.favoritesService.removeMovieFromFavorites(action.userName, action.movieId)
      .pipe(
        tap(() => {
          newState = produce(getState(), draft => {
            draft.isFetching = false;
          });
          setState(newState);
          window.location.reload();
        }),
        catchError((error: HttpErrorResponse) => {
          newState = produce(getState(), draft => {
            draft.isFetching = false;
          });
          setState(newState);
          return throwError(error);
        })
      )
      .subscribe();
  }

  @Action(UserFavouriteListMoviesResetFiltering)
  async userFavouriteListMoviesResetFiltering(
    {getState, setState}: StateContext<UserFavouriteListMoviesStateModel>) {

    let newState = produce(getState(), draft => {
      draft.isFetching = true;
    })
    setState(newState);

    newState = produce(getState(), draft => {
      draft.isFetching = false;
      draft.isFiltered = false;
      draft.movies = this.movies;
    });
    setState(newState);
  }


  @Action(UserFavouriteListMoviesReset)
  async userFavouriteListMoviesReset(
    {getState, setState}: StateContext<UserFavouriteListMoviesStateModel>) {
    setState(defaultsState);
  }
}
