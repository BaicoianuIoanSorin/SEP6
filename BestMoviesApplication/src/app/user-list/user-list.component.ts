import { Component } from '@angular/core';

@Component({
  selector: 'app-user-list',
  template: `<router-outlet></router-outlet>`,
  styleUrls: ['./user-list.component.scss']
})
export class UserListComponent {

}
