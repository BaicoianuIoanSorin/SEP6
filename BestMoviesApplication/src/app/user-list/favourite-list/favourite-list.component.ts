import { Component, OnDestroy, OnInit } from '@angular/core';
import { Select, Store } from "@ngxs/store";
import { UserTopListMoviesSelector } from "src/app/user-list/top-list/top-list.selector";
import { Observable } from "rxjs";
import { Movie } from "src/model/movie";
import { Router } from "@angular/router";
import { UserFavouriteListMoviesFetch, UserFavouriteListMoviesReset } from "src/app/user-list/favourite-list/favourite-list.actions";

@Component({
  selector: 'app-favourite-list',
  templateUrl: './favourite-list.component.html',
  styleUrls: ['./favourite-list.component.scss']
})
export class FavouriteListComponent implements OnInit, OnDestroy {
  alive = true;

  @Select(UserTopListMoviesSelector.isFetching)
  isFetching$: Observable<boolean>;

  @Select(UserTopListMoviesSelector.movies)
  movies$: Observable<Movie[]>;

  constructor(private store: Store,
              private router: Router) {
  }

  ngOnInit() {
    this.store.dispatch(new UserFavouriteListMoviesFetch());
  }

  redirectToMovieOverviewPage(movieId: number) {
    this.router.navigate([`/information/movies/${movieId}`]);
  }

  ngOnDestroy() {
    this.alive = false;
    this.store.dispatch(new UserFavouriteListMoviesReset());
  }
}
