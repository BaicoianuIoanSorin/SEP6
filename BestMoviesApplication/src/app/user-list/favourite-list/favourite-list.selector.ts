import { Selector } from "@ngxs/store";
import { UserFavouriteListMoviesState, UserFavouriteListMoviesStateModel } from "src/app/user-list/favourite-list/favourite-list.state";

export class UserFavouriteListMoviesSelector {
  @Selector([UserFavouriteListMoviesState])
  static isFetching(state: UserFavouriteListMoviesStateModel) {
    return state.isFetching;
  }

  @Selector([UserFavouriteListMoviesState])
  static movies(state: UserFavouriteListMoviesStateModel) {
    return state.movies;
  }
}
