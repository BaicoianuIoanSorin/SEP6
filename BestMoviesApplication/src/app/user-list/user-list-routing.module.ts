import { NgModule } from '@angular/core';
import { RouterModule, Routes } from '@angular/router';
import { UserListComponent } from './user-list.component';
import { TopListComponent } from "src/app/user-list/top-list/top-list.component";
import { FavouriteListComponent } from "src/app/user-list/favourite-list/favourite-list.component";

const routes: Routes = [
  {
    path: '',
    component: UserListComponent
  },
  {
    path:'top',
    component: TopListComponent,
  },
  {
    path:'favourites',
    component: FavouriteListComponent,
  }
  ];

@NgModule({
  imports: [RouterModule.forChild(routes)],
  exports: [RouterModule]
})
export class UserListRoutingModule { }
