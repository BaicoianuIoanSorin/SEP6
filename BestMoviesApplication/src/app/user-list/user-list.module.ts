import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';

import { UserListRoutingModule } from './user-list-routing.module';
import { UserListComponent } from './user-list.component';
import { TopListComponent } from './top-list/top-list.component';
import {
  NbButtonModule,
  NbCardModule,
  NbIconModule,
  NbListModule,
  NbOptionModule,
  NbSelectModule,
  NbSpinnerModule,
  NbToastrService
} from "@nebular/theme";
import { NgxsModule } from "@ngxs/store";
import { UserTopListMoviesState } from "src/app/user-list/top-list/top-list.state";
import { HttpClientModule } from "@angular/common/http";
import { AuthService } from "src/core/services/auth.service";
import { ApiService } from "src/core/services/api.service";
import { FavouriteListComponent } from './favourite-list/favourite-list.component';
import { UserFavouriteListMoviesState } from "src/app/user-list/favourite-list/favourite-list.state";
import { RouterModule } from "@angular/router";
import { SharedModule } from "src/app/shared/shared.module";

export const TOP_LIST_COMPONENTS = [
  UserListComponent,
  TopListComponent,
  FavouriteListComponent,
]

export const TOP_LIST_NEBULAR_COMPONENTS = [
  CommonModule,
  UserListRoutingModule,
  NbButtonModule,
  NbCardModule,
  NbIconModule,
  NbListModule,
  NbOptionModule,
  NbSelectModule,
  NbSpinnerModule
]

export const TOP_LIST_PROVIDERS = [
  AuthService,
  ApiService,
  NbToastrService
]

export const TOP_LIST_STATES = [
  UserTopListMoviesState,
  UserFavouriteListMoviesState,
]

export const TOP_LIST_NGXS_CONFIG = [
  NgxsModule.forFeature(TOP_LIST_STATES),
]
@NgModule({
  declarations: [
    TOP_LIST_COMPONENTS,
  ],
  imports: [
    CommonModule,
    SharedModule,
    RouterModule,
    HttpClientModule,
    TOP_LIST_NEBULAR_COMPONENTS,
    TOP_LIST_NGXS_CONFIG,
  ],
  providers: [
    TOP_LIST_PROVIDERS
  ]
})
export class UserListModule { }
