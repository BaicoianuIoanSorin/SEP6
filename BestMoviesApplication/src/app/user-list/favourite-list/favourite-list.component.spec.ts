import { createComponentFactory, Spectator, SpectatorFactory } from "@ngneat/spectator/jest";
import { CommonModule } from "@angular/common";
import { TOP_LIST_COMPONENTS, TOP_LIST_NEBULAR_COMPONENTS, TOP_LIST_PROVIDERS, TOP_LIST_STATES } from "../user-list.module";
import { HttpClientModule } from "@angular/common/http";
import { RouterTestingModule } from "@angular/router/testing";
import { NgxsModule, Store } from "@ngxs/store";
import { of } from "rxjs";
import { InjectionToken } from "@angular/core";
import { NbToastrService, NbStatusService } from "@nebular/theme";
import { ApiService } from "../../../core/services/api.service";
import { FavouriteListComponent } from "./favourite-list.component";

describe('app > user-list > favourite-list > favourite-list.component.spec.ts', () => {
  let spectator: Spectator<FavouriteListComponent>;
  let createComponent: SpectatorFactory<FavouriteListComponent> = createComponentFactory({
    component: FavouriteListComponent,
    imports: [
      CommonModule,
      HttpClientModule,
      RouterTestingModule.withRoutes([]),
      NgxsModule.forRoot(TOP_LIST_STATES),
      ...TOP_LIST_NEBULAR_COMPONENTS,
    ],
    declarations: [TOP_LIST_COMPONENTS],
    providers: [
      TOP_LIST_PROVIDERS,
      {
        provide: Store,
        useValue: {
          dispatch: jest.fn(),
          selectSnapshot: jest.fn(),
          select: jest.fn(() => of()),
        },
      },
      {
        provide: NbToastrService,
        useValue: {
          show: jest.fn(),
        },
      },
      {
        provide: NbStatusService,
        useValue: {
          show: jest.fn(),
          isCustomStatus: jest.fn(),
        },
      },
      {
        provide: ApiService,
        useValue: {
          show: jest.fn(),
        },
      },
      {
        provide: new InjectionToken('Default toastr options'),
        useValue: {},
      },
    ],
  });

  beforeEach(() => {
    spectator = createComponent();
  });

  it('should create', () => {
    expect(spectator.component).toBeTruthy();
  });

  it('should match snapshot', () => {
    spectator.detectChanges();

    // assert
    expect(spectator.fixture).toMatchSnapshot();
  });
});
