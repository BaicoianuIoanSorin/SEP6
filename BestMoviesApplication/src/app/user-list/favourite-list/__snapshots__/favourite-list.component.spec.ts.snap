// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`app > user-list > favourite-list > favourite-list.component.spec.ts should match snapshot 1`] = `
<app-favourite-list
  __isFetching$__selector={[Function Observable]}
  __isFiltered$__selector={[Function Observable]}
  __movies$__selector={[Function Observable]}
  alive={[Function Boolean]}
  authService={[Function AuthService]}
  router={[Function Router]}
  store={[Function Object]}
  user="null"
  usersFavouriteMoviesPlaceholder={[Function String]}
>
  <div
    class="favourite-list"
  >
    <div
      class="best-movies"
    >
      <nb-card>
        <nb-card-header>
          <span
            class="title"
          >
             Favourite list of movies 
          </span>
          <div
            class="right-side"
          >
            <bm-search />
          </div>
        </nb-card-header>
        <nb-card-body>
        </nb-card-body>
      </nb-card>
    </div>
  </div>
</app-favourite-list>
`;
